app/src/main/java/com/github/siela1915/bootcamp/CountDownTimerWithPause.java

```diff
@@ -0,0 +1,167 @@
+**/
+public abstract class CountDownTimerWithPause {
+
+    /**
+     * Millis since epoch when alarm should stop.
+     */
+    private final long mMillisInFuture;
+
+    /**
+     * The interval in millis that the user receives callbacks
+     */
+    private final long mCountdownInterval;
+
+    private long mStopTimeInFuture;
+
+    private long mPauseTime;
+
+    private boolean mCancelled = false;
+
+    private boolean mPaused = false;
+    private static final int MSG = 1;
+    // handles counting down
+    private Handler mHandler = new Handler() {
+        @Override
+        public void handleMessage(Message msg) {
+            synchronized (CountDownTimerWithPause.this) {
+                if (!mPaused) {
+                    final long millisLeft = mStopTimeInFuture - getRealTime();
+
+                    if (millisLeft <= 0) {
+                        onFinish();
+                    } else if (millisLeft < mCountdownInterval) {
+                        // no tick, just delay until done
+                        sendMessageDelayed(obtainMessage(MSG), millisLeft);
+                    } else {
+                        long lastTickStart = getRealTime();
+                        onTick(millisLeft);
+
+                        // take into account user's onTick taking time to execute
+                        long delay = lastTickStart + mCountdownInterval - getRealTime();
+
+                        // special case: user's onTick took more than interval to
+                        // complete, skip to next interval
+                        while (delay < 0) delay += mCountdownInterval;
+
+                        if (!mCancelled) {
+                            sendMessageDelayed(obtainMessage(MSG), delay);
+                        }
+                    }
+                }
+            }
+        }
+    };
+
+    /**
+     * @param millisInFuture The number of millis in the future from the call
+     *   to {@link #start()} until the countdown is done and {@link #onFinish()}
+     *   is called.
+     * @param countDownInterval The interval along the way to receive
+     *   {@link #onTick(long)} callbacks.
+     */
+    public CountDownTimerWithPause(long millisInFuture, long countDownInterval) {
+        mMillisInFuture = millisInFuture;
+        mCountdownInterval = countDownInterval;
+    }
+
+    public CountDownTimerWithPause(long millisInFuture, long countDownInterval, Handler handler) {
+        mMillisInFuture = millisInFuture;
+        mCountdownInterval = countDownInterval;
+        mHandler = handler;
+    }
+
+    /**
+     * Start the countdown.
+     */
+    public synchronized final void start() {
+        if (mMillisInFuture <= 0) {
+            onFinish();
+        }
+        mStopTimeInFuture = getRealTime() + mMillisInFuture;
+        mHandler.sendMessage(mHandler.obtainMessage(MSG));
+        mCancelled = false;
+        mPaused = false;
+    }
+
+    /**
+     * Pause the countdown.
+     */
+    public long pause() {
```

Comment:
Could the pause() and resume() methods cause problems if the user clicks on the corresponding buttons multiple times, with a delay? Pausing once and then once again after 5 minutes would break the timer. Consider adding checks for such cases.

Reply:
Thanks for reminding! :D I have the checking logic implemented in the TimerViewModel because I think how those edge cases are handled is open for the component that uses this timer to decide. Let me know if you have further concern on this. :)

Reply:
Oh I missed that my bad. Looks good then!